ompi/errhandler/errhandler.c

```diff
@@ -250,7 +250,7 @@ void ompi_errhandler_callback(size_t refid, pmix_status_t status,
         cbfunc(PMIX_EVENT_ACTION_COMPLETE, NULL, 0, NULL, NULL, cbdata);
     }
     /* our default action is to abort */
-    ompi_mpi_abort(MPI_COMM_WORLD, status);
```

Comment:
This is the PMIx callback. In the unlikely case it gets called, it will be called from deep inside PMIx on a PMIx thread, and possibly into a totally asynchronous manner. Basically, this makes the error reporting somewhat async, not a desirable capability from an MPI user perspective.

Reply:
There are at least 2 concerns here. First, most applications are not build in an interruptible way,  most data structures are not properly protected. As a result, having suddenly another thread potentially calling a user provided function might lead to unexpected behaviors and/or race conditions. Second, reporting an error in another thread context (a thread that is not even created by the user), is making the error handling extremely complex from the user perspective, as sha now has to cope with an async error reporting, even in the case where MPI is initialized in THREAD_SINGLE mode.

Last, but not least, even if the error handler is an OMPI callback, we might be altering the internal MPI status leading to collisions with other MPI functions that could be executing in the user thread. For the sake of our users, we should not allow any OMPI/PMIx internal threads spill out of the library and call into user space from any execution threads that are not currently in an MPI function. Instead, the errors reporting should be delayed until the user is calling an MPI function (or we could have an info key/MCA param allowing for async reporting).

Reply:
The expectation was that the MPI layer would "threadshift" this event into its own context (using whatever mechanism you folks prefer) rather than calling some internal global function from within the PMIx callback. You are certainly welcome to make that change - I believe I wrote that piece solely because (a) nobody else would do it, (b) we needed some mechanism to abort the MPI layer, and (c) I had no idea what "threadshift" mechanism you folks might want to use.

Reply:
When the callback was calling abort the behavior was different, the app would suddenly disappear with no possible feedback from the user. The change we are talking about here, change that behavior and replace it with a callback, that might have been set by the user. You are right, the correct behavior is to catch this PMIx reporting and delay it until it can be reported via one of MPI functions.

Reply:
Added this commit to the PR:
https://github.com/open-mpi/ompi/pull/7798/commits/1dbd01d0af0db3f001d207495fa8658180e7e543